api/src/main/java/org/openmrs/module/appointmentscheduling/api/AppointmentService.java

```diff
@@ -980,4 +980,38 @@ Integer getHistoryCountByConditions(Date fromDate, Date endDate,
     @Authorized(AppointmentUtils.PRIV_SCHEDULE_APPOINTMENTS)
 	Appointment bookAppointment(Appointment appointment, Boolean allowOverbook)
 			throws TimeSlotFullException;
+
+	/**
+	 * returns daily count for the statuses
+	 * scheduled
+	 * missed
+	 * completed
+	 */
+	@Authorized(AppointmentUtils.PRIV_VIEW_APPOINTMENTS)
+    List<Appointment> getDistinctDailyAppointment(Date fromDate,
+			Date toDate, Location location, Provider provider, AppointmentType appointmentType,
+			AppointmentStatus status, VisitType visitType) throws APIException;
+
+	/**
+	 * helper method that counts number of appointments per day
+	 */
+	@Authorized(AppointmentUtils.PRIV_VIEW_APPOINTMENTS)
+	Integer getDailyAppointmentsCount(Appointment appointment) throws APIException;
+
+	/**
+	 * returns list of patients and their appointments
+	 * who came late/early for their visits
+	 */
+	@Authorized(AppointmentUtils.PRIV_VIEW_APPOINTMENTS)
+	List<Appointment> getLateAndEarlyVisits(Date fromDate,
+								  Date toDate,  Location location, Provider provider, AppointmentType appointmentType,
+								  AppointmentStatus status, VisitType visitType) throws APIException;
+
```

Comment:
To follow up on this, this method (and geLateAppointments) should not include the status and visitType parameters since they aren't used int the method.